File: csharp/grade-school/GradeSchool.cs
using System;
using System.Linq;
using System.Collections.Generic;

public class GradeSchool
{
    Dictionary<int, List<string>> school = new Dictionary<int, List<string>>();

    public void Add(string student, int grade)
    {
        if(!school.ContainsKey(grade))
        {
            school.Add(grade, new List<string>());
        }
        school[grade].Add(student);
    }

    public IEnumerable<string> Roster()
    {
        return school
        .OrderBy(gradeToClassroom => gradeToClassroom.Key)
        .SelectMany( (gradeToClassroom)=> gradeToClassroom.Value.OrderBy(student => student));
    }

    public IEnumerable<string> Grade(int grade)
    {
        return school.ContainsKey(grade) ?  school[grade].OrderBy( student => student) :  Enumerable.Empty<string>();
    }
}

File: fsharp/grade-school/GradeSchool.fs
module GradeSchool

type School = Map<int, string list>

let empty: School = 
    let x: School = Map.empty<int, string list>
    x

let add (student: string) (grade: int) (school: School): School = 
    let students = (school.[grade] @ [student])
    school.Add (grade, students)
    
let roster (school: School): string list = 
    school.Keys |> Seq.cast |> Seq.map (fun k -> school.[k]) |> Seq.concat |> List.ofSeq

let grade (number: int) (school: School): string list = 
    school.[number]
